core/lib/dal/src/contract_verification_dal.rs

```diff
@@ -560,6 +583,285 @@ impl ContractVerificationDal<'_, '_> {
+            .map(|h| h.hash())
+            .unwrap_or_default();
+
+        let Some((mut info, bytecode_keccak256, bytecode_without_metadata_keccak256)) =
+            sqlx::query!(
+                r#"
+                SELECT
+                    verification_info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256
+                FROM
+                    contract_verification_info_v2
+                WHERE
+                    bytecode_keccak256 = $1
+                    OR
+                    (
+                        bytecode_without_metadata_keccak256 IS NOT null
+                        AND bytecode_without_metadata_keccak256 = $2
+                    )
+                "#,
+                identifier.bytecode_sha3.as_bytes(),
+                bytecode_without_metadata_sha3.as_bytes()
+            )
+            .try_map(|row| {
+                let info = serde_json::from_value::<VerificationInfo>(row.verification_info)
+                    .decode_column("verification_info")?;
+                let bytecode_keccak256 = H256::from_slice(&row.bytecode_keccak256);
+                let bytecode_without_metadata_keccak256 = row
+                    .bytecode_without_metadata_keccak256
+                    .map(|h| H256::from_slice(&h));
+                Ok((
+                    info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256,
+                ))
+            })
+            .instrument("get_contract_verification_info_v2#perfect_match")
+            .with_arg("address", &address)
+            .fetch_optional(self.storage)
+            .await?
+        else {
+            return Ok(None);
+        };
+        if identifier.bytecode_sha3 != bytecode_keccak256 {
+            // Sanity check
+            if bytecode_without_metadata_keccak256.is_none()
+                || identifier.bytecode_without_metadata_sha3.map(|h| h.hash())
+                    != bytecode_without_metadata_keccak256
+            {
+                tracing::error!(
+                    contract_address = %address,
+                    identifier = ?identifier,
+                    bytecode_keccak256 = %bytecode_keccak256,
+                    info = ?info,
+                    "Bogus verification info fetched for contract",
+                );
+                anyhow::bail!("Internal error: bogus verification info detected");
+            }
+
+            // Mark the contract as partial match (regardless of other issues).
+            info.verification_problems = vec![VerificationProblem::IncorrectMetadata];
+        }
+        Ok(Some(info))
+    }
+
+    /// Checks if migration from `contracts_verification_info` to `contract_verification_info_v2` is performed
+    /// by checking if the latter has more or equal number of rows.
+    pub async fn is_verification_info_migration_performed(&mut self) -> DalResult<bool> {
+        let count_v1 = sqlx::query!(
```

Comment:
Nit: Not likely an issue in practice, but it looks possible to use a single DB query here to make the query transactional (alternatively, open a transaction explicitly, but it looks like an overkill).